Allow to override parents defined hooks

josemoyab opened the pull request:
Overriding a hook in a child class currently overrides all the hooks defined by its parents. This PR aims to allow overriding only the desired hooks based on the hook class. However, it doesn’t permit adding the same hook twice on the same kind.

This PR also consider `around` hooks while checking if all hooks excluded a method

gtrias commented:
> Overriding a hook in a child class currently overrides all the parents’ defined hooks. This PR helps to override the desired ones based on the hook class.

@josemoyab can you try if the configuration works with this changes made by @atd ?

https://github.com/factorialco/captain_hook/pull/10

josemoyab commented:
> > Overriding a hook in a child class currently overrides all the parents’ defined hooks. This PR helps to override the desired ones based on the hook class.
> 
> @josemoyab can you try if the configuration works with this changes made by @atd ?
> 
> #10

I tested the configuration with the changes made by @atd, and everything worked as expected.

gtrias commented on lib/captain_hook.rb:
```diff
@@ -44,17 +44,19 @@ def hook(
     def get_hooks(kind)
-      # Only get hooks from the most specific class that defines them
-      return hooks[kind].values if hooks[kind].any?
+      hook_list = Hash.new { |h, k| h[k] = [] }
```
what is this? why it needs to be initialized in this way?

josemoyab replied:
This is a hash where we’ll store and organize the hooks defined by a class and its ancestors. The key is the hook class name, and the value is an array containing all the occurrences of that hook.

gtrias commented on lib/captain_hook.rb:
```diff
@@ -44,17 +44,19 @@ def hook(
-      # If no hooks found anywhere in the chain, return empty array
-      []
+      # Return an array containing the first element of each key in hook_list
+      hook_list.values.map(&:first)
```
why `&:first`? how this `hook_list.values` looks like?

josemoyab replied:
On each iteration of the loop, we stored all the occurrences of hooks from a class and its ancestors in the `hook_list`. Since the array is sorted, selecting the first element means that we retrieve the hook definition closer to the child class.

Considering the following class definitions:: 

```
class Child
   hook :before, hook: Hook1.new, ...
end

class Parent < Child
   hook :before, hook: Hook1.new, ...
   hook :before, hook: Hook2.new, ...
end
```

Calling `get_hooks(:before)` will set the value of `hook_list` to this:
```
{
  "Hook1": [Hook1 (from Child), Hook1 (from Parent),
  "Hook2": [Hook2 (from Parent)]
}
```

and `hook_list.values` will looks like:

```
[
[Hook1 (from Child), Hook1 (from Parent)],
[Hook2 (from Parent)]
]
```

So mapping the values using `first` will produce an array like:

```
[Hook1 (from Child), Hook2 (from Parent)]
```

gtrias reviewed (commented):
I don't fully understand the fix. What essentially changed?

gtrias commented:
> > > Overriding a hook in a child class currently overrides all the parents’ defined hooks. This PR helps to override the desired ones based on the hook class.
> > 
> > 
> > @josemoyab can you try if the configuration works with this changes made by @atd ?
> > #10
> 
> I tested the configuration with the changes made by @atd, and everything worked as expected.

You mean that merging @atd changes this PR is not needed anymore?

josemoyab commented:
> > I tested the configuration with the changes made by @atd, and everything worked as expected.
> 
> You mean that merging @atd changes this PR is not needed anymore?

It means I merge his branch and mine and everything worked as expected

josemoyab commented:
> I don't fully understand the fix. What essentially changed?

I added a comment with an example, but TL;DR I changed `get_hooks` method so if a class defines some hooks instead of returning only the [hooks defined by itself](https://github.com/factorialco/captain_hook/blob/main/lib/captain_hook.rb#L48), it also [include hooks defined by its parents](https://github.com/factorialco/captain_hook/pull/12/files#diff-1d81e155a8b74e06e72d2c9f6fc86a76e7a9f9dbe53609934ef9ab769375b989R50).

atd commented:
@josemoyab please be aware that my branch is failing in Factorial

josemoyab commented:
> @josemoyab please be aware that my branch is failing in Factorial

No worries, I've [tested this branch](https://github.com/factorialco/factorial/actions/runs/12081277961?pr=54759) in factorial and everything looks good 

@gtrias I've simplified the `get_hooks` method as we discussed

gtrias reviewed (approved):
:rocket:
